### ISSUE TITLE
question about statically linking perl

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#741]
Migrated from [rt.perl.org#1664](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=1664) (status was 'resolved')

Searchable as RT1664$

**Comment by p5pRT** [comment]
### From mlevine@reshape.com

I am using gcc\-2\.95\.1 and trying to compile perl 5\.004\_01 statically\.
I have configured it with a command like this&#8203;:

&nbsp;    \./Configure \-de \-Dprefix='/home/mlevine/reshape\.static/bin/perl ' \-Dlibs='\-lm 
&nbsp;    \-lstdc\+\+ \-lnsl ' \-Dlibpth='\-L/home/mlevine/reshape\.static/blib '
&nbsp;    \-Dldflags=' \-static ' \-Dso='none ' \-Dcc='/home2/mlevine/reshape\_tools/bin/gcc ' 
&nbsp;    \-Dcf\_email='mlevine@&#8203;reshape\.com ' \-Dmake='/home2/byron/reshape\_tools/bin/gmake ' 
&nbsp;    \-Dextensions='Reshape POSIX Fcntl IO Socket GD DProf GIFgraph DateManip ' 
&nbsp;    \-Ddynamic\_ext=' ' \-Dstatic\_ext='Reshape POSIX Fcntl IO Socket GD ' 
&nbsp;    \-Dccflags='\-DRESHAPE '

When I make perl\, I die during the link of miniperl\.  The commands being generated
to compile and link miniperl are&#8203;:

 /home2/mlevine/reshape\_tools/bin/gcc \-DPERL\_CORE \-c \-DRESHAPE \-I/usr/local/include \-O miniperlmain\.c
 /home2/mlevine/reshape\_tools/bin/gcc \-static \-o miniperl miniperlmain\.o libperl\.a \-lm \-lstdc\+\+ \-lnsl


When the make process tries to create miniperl\, it gets undefined symbols\, like this&#8203;:

&nbsp;    Undefined                 first referenced^M
&nbsp;     symbol                             in file^M
&nbsp;    dlclose                             /usr/lib/libc\.a\(nss\_deffinder\.o\)^M
&nbsp;    dlsym                               /usr/lib/libc\.a\(nss\_deffinder\.o\)^M
&nbsp;    dlopen                              /usr/lib/libc\.a\(nss\_deffinder\.o\)^M
&nbsp;    ld&#8203;: fatal&#8203;: Symbol referencing errors\. No output written to miniperl^M
&nbsp;    collect2&#8203;: ld returned 1 exit status^M

It seems like either miniperlmain\.o or libperl\.a is accessing a function in 
libc\.a which then wants to access dlcose\(\)\, dlsym\(\)\, and dlopen\(\)\, but of course\,
I have no libraries for those since I am telling the linker not to look for \.so's\.
In other words\, there is a libdl\.so on my system\, but no libdl\.a \(which makes
sense since I should want this from a dynamically linked executable\)\.

Any ideas on what I am doing wrong?

**Comment by p5pRT** [comment]
### From @gsar

On Tue\, 19 Oct 1999 08&#8203;:48&#8203;:15 PDT\, mlevine wrote&#8203;:
>I am using gcc\-2\.95\.1 and trying to compile perl 5\.004\_01 statically\.
&nbsp;                                                  ^^^^^^^^

Sorry I can't help with your real problem\, but why such an old
version?


Sarathy
gsar@&#8203;activestate\.com

**Comment by p5pRT** [comment]
### From @doughera88

On Tue\, 19 Oct 1999\, mlevine wrote&#8203;:

> Hi \-
> 
> I am using gcc\-2\.95\.1 and trying to compile perl 5\.004\_01 statically\.
> I have configured it with a command like this&#8203;:

Assuming \(from the nss\_deffinder stuff\, i\.e\. the "Name Service Switch"
stuff\) that you're on Solaris\,

>  /home2/mlevine/reshape\_tools/bin/gcc \-static \-o miniperl
> miniperlmain\.o libperl\.a \-lm \-lstdc\+\+ \-lnsl

>     Undefined                 first referenced^M
>      symbol                             in file^M
>     dlopen                              /usr/lib/libc\.a\(nss\_deffinder\.o\)^M
 
> It seems like either miniperlmain\.o or libperl\.a is accessing a
> function in libc\.a which then wants to access dlcose\(\)\, dlsym\(\)\, and
> dlopen\(\)\, but of course\, I have no libraries for those since I am
> telling the linker not to look for \.so's\. In other words\, there is a
> libdl\.so on my system\, but no libdl\.a \(which makes sense since I
> should want this from a dynamically linked executable\)\.

This diagnosis looks reasonable\.

> Any ideas on what I am doing wrong?

I don't think you're doing anything "wrong"\, but you may have to rethink
your quest to have \*everything\* statically linked\.  From the
nsswitch\.conf\(4\) man page \[trimmed\]

FILES

&nbsp;     A source named SSS is implemented by a shared  object  named
&nbsp;     nss\_SSS\.so\.1 that resides in /usr/lib\.

&nbsp;     /usr/lib/nss\_compat\.so\.1 implements "compat" source
&nbsp;     /usr/lib/nss\_dns\.so\.1    implements "dns" source
&nbsp;     /usr/lib/nss\_files\.so\.1  implements "files" source
&nbsp;     /usr/lib/nss\_nis\.so\.1    implements "nis" source

I don't think you'll easily be able to build a perl on Solaris that uses
name services without allowing some shared objects to sneak into the
picture somehow\.  Whether or not this is a problem depends ultimately on
just why you want to build everything static\.

Hope this helps\,

&nbsp;    Andy Dougherty        doughera@&#8203;lafayette\.edu
&nbsp;    Dept\. of Physics
&nbsp;    Lafayette College\, Easton PA 18042

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Mlevine \<mlevine@&#8203;reshape\.com> writes&#8203;:
>Hi \-
>
>I am using gcc\-2\.95\.1 and trying to compile perl 5\.004\_01 statically\.
>I have configured it with a command like this&#8203;:
>
>    \./Configure \-de \-Dprefix='/home/mlevine/reshape\.static/bin/perl ' \-Dlibs='\-lm 
>    \-lstdc\+\+ \-lnsl ' \-Dlibpth='\-L/home/mlevine/reshape\.static/blib '
>    \-Dldflags=' \-static ' \-Dso='none ' \-Dcc='/home2/mlevine/reshape\_tools/bin/gcc ' 
>    \-Dcf\_email='mlevine@&#8203;reshape\.com ' \-Dmake='/home2/byron/reshape\_tools/bin/gmake ' 
>    \-Dextensions='Reshape POSIX Fcntl IO Socket GD DProf GIFgraph DateManip ' 
>    \-Ddynamic\_ext=' ' \-Dstatic\_ext='Reshape POSIX Fcntl IO Socket GD ' 
>    \-Dccflags='\-DRESHAPE '

I cannot see anything in there which tells us which OS you are using\.

>
>When I make perl\, I die during the link of miniperl\.  The commands being generated
>to compile and link miniperl are&#8203;:
>
> /home2/mlevine/reshape\_tools/bin/gcc \-DPERL\_CORE \-c \-DRESHAPE \-I/usr/local/include \-O miniperlmain\.c
> /home2/mlevine/reshape\_tools/bin/gcc \-static \-o miniperl miniperlmain\.o libperl\.a \-lm \-lstdc\+\+ \-lnsl
>
>
>When the make process tries to create miniperl\, it gets undefined symbols\, like this&#8203;:
>
>    Undefined                 first referenced^M
>     symbol                             in file^M
>    dlclose                             /usr/lib/libc\.a\(nss\_deffinder\.o\)^M
>    dlsym                               /usr/lib/libc\.a\(nss\_deffinder\.o\)^M
>    dlopen                              /usr/lib/libc\.a\(nss\_deffinder\.o\)^M
>    ld&#8203;: fatal&#8203;: Symbol referencing errors\. No output written to miniperl^M
>    collect2&#8203;: ld returned 1 exit status^M

If this is Solaris you may just not be able to build fully static\.
You might try moving \-lnsl up\-front 

>
>It seems like either miniperlmain\.o or libperl\.a is accessing a function in 
>libc\.a which then wants to access dlcose\(\)\, dlsym\(\)\, and dlopen\(\)\, but of course\,
>I have no libraries for those since I am telling the linker not to look for \.so's\.
>In other words\, there is a libdl\.so on my system\, but no libdl\.a \(which makes
>sense since I should want this from a dynamically linked executable\)\.

If libc\.a needs dlopen\(\) & co\. you are in bad shape\.
But presumably it is only "name service switch" code that needs
this stuff\. Hence the suggestion about \-lnsl above\.

But note that this means you need to specify HOW names are going 
to get resolved \(FILES\, NIS\, DNS\) at link time\. This is usually 
a bad idea and limits where executable can be used\. 

>
>Any ideas on what I am doing wrong?

Why do you want an all static version?

>
>
>
>
>
>
>
>Message from the perl bug squashing team at 'perlbug@&#8203;perl\.org'
\-\- 
Nick Ing\-Simmons
